Add backwards compatibility for the scraper (#7664)
sunpy/sunpy: docs/topic_guide/extending_fido.rst

* A class method :meth:`~sunpy.net.base_client.BaseClient.register_values`; this registers the "attrs" that are supported by the client.
  It returns a dictionary where keys are the supported attrs and values are lists of tuples.
  Each `tuple` contains the "attr" value and its description.
* A class attribute ``pattern``; this is a string used to match all URLs supported by the client and extract necessary metadata from the matched URLs.

>   The time and other metadata attributes for extraction are written within double curly-braces ``{{}}`` in a `parse` format. Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``. An example of how such a pattern looks like is given in the algorithm explanation below.

Reviewer: ```suggestion
  The time and other metadata attributes for extraction are written within double curly-braces ``{{}}`` in a `parse` format.
  Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``.
  An example of how such a pattern looks like is given in the algorithm explanation below.
```

* A class method :meth:`~sunpy.net.base_client.BaseClient.register_values`; this registers the "attrs" that are supported by the client.
  It returns a dictionary where keys are the supported attrs and values are lists of tuples.
  Each `tuple` contains the "attr" value and its description.
* A class attribute ``pattern``; this is a string used to match all URLs supported by the client and extract necessary metadata from the matched URLs.
  The time and other metadata attributes for extraction are written within double curly-braces ``{{}}`` in a `parse <https://github.com/r1chardj0n3s/parse/>`__ format.
  Regular placeholders for Python format strings can still be included via single curly braces ``{}`` with their respective parameters passed as the ``kwargs``.
  An example of how such a pattern looks like is given in the algorithm explanation below.

> Note: The scraper has supported regex-based patterns for a long time, which is what it will still expect by default. However that is in the process of being replaced with parse-style patterns in the future versions of the package. To use the newer parse-style patterns, it's currently required to pass it as the ``format`` argument to the scraper.

Reviewer: ```suggestion
Note: The scraper has supported regex-based patterns for a long time, which is what it will still expect by default.
However that is in the process of being replaced with parse-style patterns in the future versions of the package.
To use the newer parse-style patterns, it's currently required to pass it as the ``format`` argument to the scraper.
```

Reviewer: We want to enable the new format by default and raise a warning if its the old way.

Each such client relies on the `~sunpy.net.scraper.Scraper` to be able to query for files using the :meth:`~sunpy.net.scraper.Scraper.filelist` method. The general algorithm to explain how the `~sunpy.net.scraper.Scraper` is able to do this is:
A brief explanation of how the Scraper works is as follows:

1. Drop the filename from the pattern and generate a list of directories to search for files.
2. For each directory, get a list of files present in it.
3. For each file, check if it matches the pattern.
4. For each file that matches the pattern, check if it is in the timerange. If it is, add it to the output.

For a more verbose in-depth explanation on how this is accomplished internally:

> 1. A Scraper object takes as input the generalised ``pattern`` of how a desired filepath looks like, in the ``parse`` format. Upon initialisation, a version of the pattern following the datetime format is also internally generated, called the ``dt_pattern``.

Reviewer: ```suggestion
1. The Scraper takes as input the generalised ``pattern`` of how a desired file path looks like, in the ``parse`` format.
   Upon initialisation, a version of the pattern following the datetime format is also internally generated, called the ``dt_pattern``.
```

.. code-block:: python

    >>> from sunpy.net import Scraper
    >>> pattern = ('http://proba2.oma.be/{instrument}/data/bsd/{{year:4d}}/{{month:2d}}/{{day:2d}}/'
    ...            '{instrument}_lv1_{{year:4d}}{{month:2d}}{{day:2d}}_{{hour:2d}}{{minute:2d}}{{second:2d}}.fits')
    >>> s = Scraper(format=pattern, instrument='swap')

>     >>> s.dt_pattern

Reviewer: Thoughts on renaming this to "datetiime_pattern"?

Author: i think that works better, yeah

    'http://proba2.oma.be/swap/data/bsd/%Y/%m/%d/swap_lv1_%Y%m%d_%H%M%S.fits'

> The smallest unit of time / time-step for that directory-pattern (the full ``dt_pattern`` except the filename at the end) is then internally detected from ``dt_pattern`` by using the :meth:`~sunpy.net.scraper_utils.extract_timestep` util.

Reviewer: ```suggestion
The smallest unit of time / time-step for that directory-pattern (the full ``dt_pattern`` except the filename at the end) is then internally detected from ``dt_pattern`` by using :meth:`~sunpy.net.scraper_utils.extract_timestep`.
```

    'http://proba2.oma.be/swap/data/bsd/2015/01/02/',
    'http://proba2.oma.be/swap/data/bsd/2015/01/03/']

> 2. The location given by the filled pattern is visited and a list of files at the location is obtained. This is handled differently depending on whether the pattern is a web URL or a ``file://`` or an ``ftp://`` path in the :meth:`~sunpy.net.scraper.Scraper.filelist` method.
> 3. Each filename is then parsed against the remaining portion of the pattern to determine if it matches.
> 4. Each such file is then checked for lying in the intended timerange using the :meth:`~sunpy.net.scraper._check_timerange` method which in turn uses :meth:`sunpy.net.scraper_utils.get_timerange_from_exdict` to get the covered timerange for each file. The files that satisfy these conditions are then added to the output.

Reviewer: ```suggestion
The location given by the filled pattern is visited and a list of files at the location is obtained.
This is handled differently depending on whether the pattern is a web URL or a ``file://`` or an ``ftp://`` path by :meth:`~sunpy.net.scraper.Scraper.filelist`.

Each filename is then parsed against the remaining portion of the pattern to determine if it matches.
Each such file is then checked for lying in the intended timerange using the :meth:`~sunpy.net.scraper._check_timerange` method which in turn uses :meth:`sunpy.net.scraper_utils.get_timerange_from_exdict` to get the covered timerange for each file.
The files that satisfy these conditions are then added to the output.
```

Suppose any file of a data archive can be described by this URL ``https://some-domain.com/%Y/%m/%d/satname_{SatelliteNumber}_{Level}_%y%m%d%H%M%S_{any-2-digit-number}.fits``:

> The new format ``pattern`` becomes ``r'https://some-domain.com{{year:4d}}/{{month:2d}}{{day:2d}}/satname_{SatelliteNumber:2d}_{Level:1d}_{{year:2d}}{{month:2d}}{{day:2d}}{{hour:2d}}{{minute:2d}}{{second:2d}}_{{:2d}}.fits'``.
> The date-time values and any other metadata attributes that we wish to extract are written within double curly-braces ``{{}}``. These metadata attributes are the desired keys for the returned dictionary and they should match with the ``attr.__name__``.
> Note that parts of such attributes can accordingly be omitted to match parts of the filename which are dynamic but not needed to be extracted. For example, ``{{:2d}}`` is used in the above example to match any 2-digit number in the filename. Similarly ``{{}}`` can be used to match a string of any length starting from its position in the filename.

Reviewer: ```suggestion
The new format ``pattern`` becomes ``r'https://some-domain.com{{year:4d}}/{{month:2d}}{{day:2d}}/satname_{SatelliteNumber:2d}_{Level:1d}_{{year:2d}}{{month:2d}}{{day:2d}}{{hour:2d}}{{minute:2d}}{{second:2d}}_{{:2d}}.fits'``.
The datetime values and any other metadata attributes that we wish to extract are written within double curly-braces ``{{}}``.
These metadata attributes are the desired keys for the returned dictionary and they should match with the ``attr.__name__``.
Note that parts of such attributes can accordingly be omitted to match parts of the filename which are dynamic but not needed to be extracted.
For example, ``{{:2d}}`` is used in the above example to match any 2-digit number in the filename.
Similarly ``{{}}`` can be used to match a string of any length starting from its position in the filename.
```

        return adict

> 

Reviewer: ```suggestion
```
